### PR TITLE
Calibrations release 1.4

### PR DESCRIPTION
**Pull request by ethanandrews**
Final release version for v1.4.0. Initial suite of Calibration functionality.

### CODE REVIEW

File: roundabout/templates/inventory/ajax_inventory_detail.html [hunk 1]
```diff
@@ -143,6 +143,25 @@
...
+                    data-location-id="{{ inventory_item.location.id}}"
+                    data-node-type="calibration_event">Add Calibration Coefficients</a>
+
+                <a  id="add_conf_const_action"
+                    class="dropdown-item" 
+                    href="#"
+                    data-detail-url="{% url 'configs_constants:config_event_value_add' inventory_item.id %}"
+                    data-node-id="{{ inventory_item.id }}"
+                    data-location-id="{{ inventory_item.location.id}}"
+                    data-node-type="config_event}">Add Configuration/Constant Event</a>
```
Comment (spetillo) [inline roundabout/templates/inventory/ajax_inventory_detail.html:161]: Nit: Suggest `Add Configuration/Constant Event` --> `Add Configurations/Constants` to use consistent cording with `Add Calibration Coefficients` above.

File: roundabout/calibrations/forms.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        for name in from_part.coefficient_names.all():
+            CoefficientName.objects.create(
+                calibration_name = name.calibration_name,
+                value_set_type = name.value_set_type,
+                sigfig_override = name.sigfig_override,
+                part = to_part
+            )
+
+# Validator for Part Calibration Copy
+# Checks for duplicate Coefficient Names between Parts
```
[the remark this answers is not in the document]
Comment (MarioCarloni) [reply roundabout/calibrations/forms.py:385]: When a Part is selected, from which to copy Calibration Names into another Part, the function checks if duplicate Names exist between the two Parts in question.

File: roundabout/configs_constants/migrations/0002_auto_20200609_1918.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    dependencies = [
+        ('configs_constants', '0001_initial'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='configname',
+            name='config_type',
+            field=models.CharField(choices=[('cnst', 'Constant'), ('conf', 'Configuration')], default='ct', max_length=4),
```
[the remark this answers is not in the document]
Comment (MarioCarloni) [reply roundabout/configs_constants/migrations/0002_auto_20200609_1918.py:16]: This shouldn't cause an issue. "ct" was my original representation for "Constant", but I'd updated it to "cnst" which should be found in later migrations.

File: roundabout/configs_constants/models.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    def __str__(self):
+        return self.name
+    def get_object_type(self):
+        return 'config_name'
+    CONFIG_TYPE = (
+        ("cnst", "Constant"),
+        ("conf", "Configuration"),
+    )
+    name = models.CharField(max_length=255, unique=False, db_index=True)
+    config_type = models.CharField(max_length=4, choices=CONFIG_TYPE, null=False, blank=False, default="ct")
```
[the remark this answers is not in the document]
Comment (MarioCarloni) [reply roundabout/configs_constants/models.py:51]: Django models will default to the first "choice" (Constant) from its "choices" attribute, even though the default string is incorrect. I'd changed the default type in a later commit/migration from my sub-branch. This shouldn't cause an issue on Staging now, and will be the correct default "cnst" once the Constant Default and Constant Event features are merged in.

File: roundabout/configs_constants/views.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+    def form_invalid(self, form, config_event_value_form):
+        if self.request.is_ajax():
+            if form.errors:
+                data = form.errors
+                return JsonResponse(
+                    data, 
+                    status=400
+                )
+            if config_event_value_form.errors:
+                print('conf_const updateform errors')
```
[the remark this answers is not in the document]
Comment (MarioCarloni) [reply roundabout/configs_constants/views.py:167]: I believe those print statements can be removed. They were originally to help me differentiate Event formset errors coming from either the Create or Update Classes which handle template form rendering. 'updateform errors' is shorthand for the Constant Event formset error-set which populates within the Class (ConfigEventValueUpdate) which handles updating Constant/Configuration Events.

File: roundabout/calibrations/forms.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (spetillo) [reply roundabout/calibrations/forms.py:385]: Ah, ok, that makes sense. Sounds good.

File: roundabout/calibrations/models.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        ordering = ['calibration_name']
+        unique_together = ['part','calibration_name']
+    def __str__(self):
+        return self.calibration_name
+    def get_object_type(self):
+        return 'coefficient_name'
+    VALUE_SET_TYPE = (
+        ("sl", "Single"),
+        ("1d", "1-Dimensional Array"),
+        ("2d", "2-DImensional Array"),
```
[the remark this answers is not in the document]
Comment (MarioCarloni) [reply roundabout/calibrations/models.py:46]: There's an existing migration in my sub-branch which fixes this issue. This item should be resolved once the Constant Default changes are merged.

File: roundabout/configs_constants/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (spetillo) [reply roundabout/configs_constants/views.py]: Pushed changes look good.

File: roundabout/templates/inventory/ajax_inventory_detail.html (hunk 1 above)
Comment (spetillo) [reply roundabout/templates/inventory/ajax_inventory_detail.html]: Pushed changes look good.

**Comment by spetillo** [comment]
Update: @ethanandrews  I can easily update my header script to ignore any folder(s) or specific files, so I can definitely do this automatically if you tell me what to ignore. I made a first pass [in this branch](https://github.com/WHOIGit/ooicgsn-roundabout/compare/calibrations-release-1.4...add_headers_to_v1.4?expand=1), just ignoring the stuff in the `static` folder. If there's anything else I should ignore, I can re-do that branch with more stuff excluded and will put in a PR to the release branch first thing tomorrow (Fri).

**Comment by ethanandrews** [comment]
@spetillo Yes, you've got it right at the moment. You can safely ignore everything in `roundabout/static`

There's a `vendor` directory under the `js` directory in there as well, but I need to reorganize that folder again as we move stuff off of CDNs and make everything local

**Comment by spetillo** [comment]
Ok, sounds like we can just take the branch I made as-is, then. PR to pull into this branch: https://github.com/WHOIGit/ooicgsn-roundabout/pull/104

We should be good to release after that is approved and pulled in. Woo hoo!

**Pull request merged**